src/v/cloud_storage/topic_recovery_service.cc

```diff
@@ -0,0 +1,591 @@
+  , _remote{remote}
+  , _topic_state{topic_state}
+  , _topics_frontend{topics_frontend}
+  , _topic_recovery_status_frontend{topic_recovery_status_frontend} {}
+
+ss::future<> topic_recovery_service::stop() {
+    vlog(cst_log.debug, "stopping topic recovery service");
+    co_await _gate.close();
+    vlog(cst_log.debug, "stopped topic recovery service");
+}
+
+ss::future<> topic_recovery_service::shutdown_recovery() {
+    vlog(cst_log.debug, "shutting down recovery if active");
+    if (_pending_status_timer.cancel()) {
+        vlog(cst_log.warn, "cancelled active pending status timer");
+    }
+    _as.request_abort();
+    co_return;
+}
+
+init_recovery_result
+topic_recovery_service::start_recovery(ss::httpd::request req) {
+    try {
+        if (is_active()) {
+            vlog(cst_log.warn, "A recovery is already active");
+            return {
+              .status_code = ss::httpd::reply::status_type::conflict,
+              .message = "A recovery is already active"};
+        }
+
+        recovery_request request{std::move(req)};
+        ssx::spawn_with_gate(_gate, [this, r = std::move(request)]() mutable {
+            return start_bg_recovery_task(std::move(r)).then([](auto result) {
+                if (result.has_error()) {
+                    vlog(
+                      cst_log.error,
+                      "topic recovery completed with error: {}",
+                      result.error().context);
+                } else {
+                    vlog(
+                      cst_log.info,
+                      "topic recovery completed. Pending downloads");
+                }
+            });
+        });
+        return {
+          .status_code = ss::httpd::reply::status_type::accepted,
+          .message = "recovery started"};
+    } catch (const bad_request& ex) {
+        return {
+          .status_code = ss::httpd::reply::status_type::bad_request,
+          .message = fmt::format(
+            "bad recovery request payload: {}", ex.what())};
+    } catch (const std::exception& ex) {
+        return {
+          .status_code = ss::httpd::reply::status_type::internal_server_error,
+          .message = fmt::format(
+            "recovery init failed with error: {}", ex.what())};
+    }
+}
+
+ss::future<> topic_recovery_service::propagate_state(state s) {
+    return container().invoke_on_all([s](auto& svc) {
+        svc._state = s;
+        return ss::make_ready_future<>();
+    });
+}
+
+ss::future<result<void, recovery_error_ctx>>
+topic_recovery_service::start_bg_recovery_task(recovery_request request) {
+    vlog(cst_log.info, "Starting recovery task with request: {}", request);
+    // Setting this state here ensures that another request coming in soon after
+    // on the same shard is rejected. If we set this state after the RPC call,
+    // it is possible that two requests sent back to back on the controller
+    // leader assume that no other recovery is running.
+    _state = state::starting;
```

Comment:
Since there's a scheduling point between when we last checked if a recovery was active, should we check the state first and exit if anything else is running?

Reply:
added a check